File: clients/python_pypy/model/unit_order.py
from model.action_order import ActionOrder
from model.vec2 import Vec2
from stream_wrapper import StreamWrapper
from typing import Optional

class UnitOrder:
    """Order for specific unit"""

    __slots__ = ("target_velocity","target_direction","action",)

    target_velocity: Vec2
    target_direction: Vec2
    action: Optional[ActionOrder]

    def __init__(self, target_velocity: Vec2, target_direction: Vec2, action: Optional[ActionOrder]):
        self.target_velocity = target_velocity
        """Target moving velocity"""
        self.target_direction = target_direction
        """Target view direction (vector length doesn't matter)"""
        self.action = action
        """Order to perform an action, or None"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "UnitOrder":
        """Read UnitOrder from input stream
        """
        target_velocity = Vec2.read_from(stream)
        target_direction = Vec2.read_from(stream)
        if stream.read_bool():
            action = ActionOrder.read_from(stream)
        else:
            action = None
        return UnitOrder(target_velocity, target_direction, action)
    
    def write_to(self, stream: StreamWrapper):
        """Write UnitOrder to output stream
        """
        self.target_velocity.write_to(stream)
        self.target_direction.write_to(stream)
        if self.action is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.action.write_to(stream)
    
    def __repr__(self):
        return "UnitOrder(" + \
            repr(self.target_velocity) + \
            ", " + \
            repr(self.target_direction) + \
            ", " + \
            repr(self.action) + \
            ")"

File: clients/cpp/model/UnitOrder.cpp
#include "UnitOrder.hpp"

namespace model {

UnitOrder::UnitOrder(model::Vec2 targetVelocity, model::Vec2 targetDirection, std::optional<std::shared_ptr<model::ActionOrder>> action) : targetVelocity(targetVelocity), targetDirection(targetDirection), action(action) { }

// Read UnitOrder from input stream
UnitOrder UnitOrder::readFrom(InputStream& stream) {
    model::Vec2 targetVelocity = model::Vec2::readFrom(stream);
    model::Vec2 targetDirection = model::Vec2::readFrom(stream);
    std::optional<std::shared_ptr<model::ActionOrder>> action = std::optional<std::shared_ptr<model::ActionOrder>>();
    if (stream.readBool()) {
        std::shared_ptr<model::ActionOrder> actionValue = model::ActionOrder::readFrom(stream);
        action.emplace(actionValue);
    }
    return UnitOrder(targetVelocity, targetDirection, action);
}

// Write UnitOrder to output stream
void UnitOrder::writeTo(OutputStream& stream) const {
    targetVelocity.writeTo(stream);
    targetDirection.writeTo(stream);
    if (action) {
        stream.write(true);
        const std::shared_ptr<model::ActionOrder>& actionValue = *action;
        actionValue->writeTo(stream);
    } else {
        stream.write(false);
    }
}

// Get string representation of UnitOrder
std::string UnitOrder::toString() const {
    std::stringstream ss;
    ss << "UnitOrder { ";
    ss << "targetVelocity: ";
    ss << targetVelocity.toString();
    ss << ", ";
    ss << "targetDirection: ";
    ss << targetDirection.toString();
    ss << ", ";
    ss << "action: ";
    if (action) {
        const std::shared_ptr<model::ActionOrder>& actionValue = *action;
        ss << actionValue->toString();
    } else {
        ss << "none";
    }
    ss << " }";
    return ss.str();
}

}